File: backend/app/semantic_search.py
import json
import logging
import math
import re
from pathlib import Path

from app.config import DATA_DIR, USE_TRANSFORMER_MODEL

logger = logging.getLogger("memegpt.semantic_search")

DIM = 384
_embeddings_cache: dict[str, list[float]] | None = None
_st_model = None
_model_attempted = False


def _get_transformer_model():
    global _st_model, _model_attempted
    if not USE_TRANSFORMER_MODEL or _model_attempted:
        return _st_model

    _model_attempted = True
    try:
        from sentence_transformers import SentenceTransformer
        logger.info("Loading sentence-transformers model (all-MiniLM-L6-v2)...")
        _st_model = SentenceTransformer("all-MiniLM-L6-v2")
        logger.info("SentenceTransformer model loaded successfully.")
    except Exception as e:
        logger.warning(f"Could not load SentenceTransformer ({e}). Falling back to TF-IDF semantic search.")
        _st_model = None
    return _st_model


def _load_embeddings() -> dict[str, list[float]]:
    global _embeddings_cache
    if _embeddings_cache is not None:
        return _embeddings_cache

    path = DATA_DIR / "embeddings.json"
    _embeddings_cache = {}
    if path.exists():
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
            for item in data:
                _embeddings_cache[item["id"]] = item["vector"]
            logger.info(f"Loaded {len(_embeddings_cache)} pre-computed embeddings from {path.name}")
        except Exception as e:
            logger.error(f"Error loading embeddings file: {e}")
    return _embeddings_cache


def tokenize(text: str) -> list[str]:
    return [t for t in re.sub(r"[^\w\s%]", " ", text.lower()).split() if len(t) > 1]


def _hash_term(term: str) -> int:
    h = 0
    for ch in term:
        h = (h * 31 + ord(ch)) & 0xFFFFFFFF
    return h % DIM


def _build_idf(corpus: list[list[str]]) -> dict[str, float]:
    df: dict[str, int] = {}
    n = len(corpus)
    for doc in corpus:
        for term in set(doc):
            df[term] = df.get(term, 0) + 1
    return {term: math.log((n + 1) / (count + 1)) + 1 for term, count in df.items()}


def _normalize(vec: list[float]) -> list[float]:
    mag = math.sqrt(sum(v * v for v in vec)) or 1.0
    return [v / mag for v in vec]


def _build_vector(tokens: list[str], idf: dict[str, float]) -> list[float]:
    if not tokens:
        return [0.0] * DIM
    tf: dict[str, int] = {}
    for t in tokens:
        tf[t] = tf.get(t, 0) + 1
    vec = [0.0] * DIM
    for term, count in tf.items():
        vec[_hash_term(term)] += (count / len(tokens)) * idf.get(term, 0)
    return _normalize(vec)


def _cosine(a: list[float], b: list[float]) -> float:
    n = min(len(a), len(b))
    if n == 0:
        return 0.0
    return sum(a[i] * b[i] for i in range(n))


def embed_text(text: str) -> list[float]:
    """Generate 384-dimensional L2-normalized vector for given text."""
    model = _get_transformer_model()
    if model is not None:
        try:
            vec = model.encode(text, convert_to_numpy=True).tolist()
            return _normalize(vec)
        except Exception:
            pass
    tokens = tokenize(text)
    idf = {t: 1.0 for t in tokens}
    return _build_vector(tokens, idf)



def semantic_scores(query: str, memes: list[dict]) -> dict[str, float]:
    query_tokens = tokenize(query)
    corpus = [
        tokenize(f"{m['name']} {m['dialogue']} {m['explanation']} {' '.join(m.get('keywords', []))}")
        for m in memes
    ]
    idf = _build_idf(corpus)
    query_vec = _build_vector(query_tokens, idf)
    stored = _load_embeddings()

    # Check if sentence-transformers model is available for live query embedding
    model = _get_transformer_model()
    st_query_vec = None
    if model is not None:
        try:
            st_query_vec = model.encode(query, convert_to_numpy=True).tolist()
        except Exception as e:
            logger.warning(f"Failed to encode query with model: {e}")

    scores: dict[str, float] = {}

    for i, meme in enumerate(memes):
        score = 0.0
        meme_vec = _build_vector(corpus[i], idf)
        
        # Base TF-IDF Cosine
        score += _cosine(query_vec, meme_vec) * 0.4

        # Keyword and term overlap bonus
        meme_terms = set(corpus[i])
        overlap = sum(1 for qt in query_tokens if qt in meme_terms)
        for kw in meme.get("keywords", []):
            kl = kw.lower()
            for qt in query_tokens:
                if qt in kl or kl in qt:
                    overlap += 0.5
        score += (overlap / max(len(query_tokens), 1)) * 0.35

        # Vector comparison (using stored embeddings or transformer model)
        meme_id = meme["id"]
        if meme_id in stored:
            target_vec = stored[meme_id]
            if st_query_vec is not None:
                score += _cosine(st_query_vec, target_vec) * 0.25
            else:
                score += _cosine(query_vec[: len(target_vec)], target_vec) * 0.25

        scores[meme_id] = min(score, 1.0)

    return scores


async def async_semantic_scores(query: str, memes: list[dict]) -> dict[str, float]:
    """Non-blocking async wrapper that runs CPU-heavy model inference in a threadpool."""
    from starlette.concurrency import run_in_threadpool
    return await run_in_threadpool(semantic_scores, query, memes)

File: server/src/lib/semantic-search.ts
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import type { MemeRecord } from "./types.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const EMBEDDINGS_PATH = path.resolve(__dirname, "../../data/embeddings.json");

interface EmbeddingEntry {
  id: string;
  vector: number[];
}

let cachedEmbeddings: Map<string, number[]> | null = null;

function loadEmbeddings(): Map<string, number[]> | null {
  if (cachedEmbeddings) return cachedEmbeddings;

  try {
    if (!fs.existsSync(EMBEDDINGS_PATH)) return null;
    const raw = JSON.parse(fs.readFileSync(EMBEDDINGS_PATH, "utf-8")) as EmbeddingEntry[];
    cachedEmbeddings = new Map(raw.map((e) => [e.id, e.vector]));
    return cachedEmbeddings;
  } catch {
    return null;
  }
}

function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^\w\s%]/g, " ")
    .split(/\s+/)
    .filter((t) => t.length > 1);
}

function buildTfIdfVector(tokens: string[], idf: Map<string, number>, dim: number): number[] {
  const tf = new Map<string, number>();
  for (const t of tokens) {
    tf.set(t, (tf.get(t) ?? 0) + 1);
  }

  const vec = new Array(dim).fill(0);
  let idx = 0;
  for (const [term, count] of tf) {
    const idfVal = idf.get(term) ?? 0;
    const hash = hashTerm(term, dim);
    vec[hash] += (count / tokens.length) * idfVal;
    idx++;
  }

  return normalize(vec);
}

function hashTerm(term: string, dim: number): number {
  let h = 0;
  for (let i = 0; i < term.length; i++) {
    h = (h * 31 + term.charCodeAt(i)) >>> 0;
  }
  return h % dim;
}

function normalize(vec: number[]): number[] {
  const mag = Math.sqrt(vec.reduce((s, v) => s + v * v, 0)) || 1;
  return vec.map((v) => v / mag);
}

function cosineSimilarity(a: number[], b: number[]): number {
  const len = Math.min(a.length, b.length);
  let dot = 0;
  for (let i = 0; i < len; i++) {
    dot += a[i] * b[i];
  }
  return dot;
}

function buildIdf(corpus: string[][]): Map<string, number> {
  const df = new Map<string, number>();
  const n = corpus.length;

  for (const doc of corpus) {
    const unique = new Set(doc);
    for (const term of unique) {
      df.set(term, (df.get(term) ?? 0) + 1);
    }
  }

  const idf = new Map<string, number>();
  for (const [term, count] of df) {
    idf.set(term, Math.log((n + 1) / (count + 1)) + 1);
  }
  return idf;
}

const DIM = 384;

export function semanticSearch(
  query: string,
  memes: MemeRecord[],
  limit = 15
): Map<string, number> {
  const scores = new Map<string, number>();
  const queryTokens = tokenize(query);

  const corpus = memes.map((m) =>
    tokenize([m.name, m.dialogue, m.explanation, ...m.keywords].join(" "))
  );
  const idf = buildIdf(corpus);
  const queryVec = buildTfIdfVector(queryTokens, idf, DIM);

  const embeddings = loadEmbeddings();

  for (let i = 0; i < memes.length; i++) {
    const meme = memes[i];
    let score = 0;

    // TF-IDF cosine similarity
    const memeVec = buildTfIdfVector(corpus[i], idf, DIM);
    score += cosineSimilarity(queryVec, memeVec) * 0.4;

    // Keyword overlap (BM25-lite)
    const memeTerms = new Set(corpus[i]);
    let overlap = 0;
    for (const qt of queryTokens) {
      if (memeTerms.has(qt)) overlap++;
      for (const kw of meme.keywords) {
        if (qt.includes(kw.toLowerCase()) || kw.toLowerCase().includes(qt)) {
          overlap += 0.5;
        }
      }
    }
    score += (overlap / Math.max(queryTokens.length, 1)) * 0.35;

    // Precomputed sentence-transformer embeddings (when available)
    if (embeddings?.has(meme.id)) {
      const stored = embeddings.get(meme.id)!;
      // Approximate query embedding by averaging nearest keyword-matched meme vectors
      // For proper query embedding, run scripts/generate_embeddings.py with --query mode
      // Here we boost memes whose stored vectors align with TF-IDF proxy
      const proxySim = cosineSimilarity(queryVec.slice(0, stored.length), stored);
      score += proxySim * 0.25;
    }

    scores.set(meme.id, Math.min(score, 1));
  }

  return scores;
}

export function getTopSemanticMatches(
  query: string,
  memes: MemeRecord[],
  limit: number
): { id: string; score: number }[] {
  const scores = semanticSearch(query, memes, limit);
  return [...scores.entries()]
    .map(([id, score]) => ({ id, score }))
    .sort((a, b) => b.score - a.score)
    .slice(0, limit);
}
